### PR TITLE
[SimplifyCFG] Supporting hoisting/sinking callbases with differing attrs

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[SimplifyCFG] Add tests for hoisting/sinking callbases with differing attrs; NFC**
- **[SimplifyCFG] Supporting hoisting/sinking callbases with differing attrs**

Some (many) attributes can safely be dropped to enable sinking. For
example removing `nonnull` on a return/param can't affect correctness.

NB: Related to #91101

### CODE REVIEW

File: llvm/test/Transforms/SimplifyCFG/sink-cb-diff-attrs.ll [hunk 1]
```diff
@@ -0,0 +1,219 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --function-signature --check-attributes
```
Comment (dtcxzyw) [inline llvm/test/Transforms/SimplifyCFG/sink-cb-diff-attrs.ll:1]: Missing check lines for attributes. Can you add `--check-globals`?

**Review by dtcxzyw — COMMENTED** [review]
IR diff looks good. But we need to avoid redundant copies/computations to reduce the compilation time overhead.

**Comment by goldsteinn** [comment]
> IR diff looks good. But we need to avoid redundant copies/computations to reduce the compilation time overhead.

http://llvm-compile-time-tracker.com/compare.php?from=f7c3309b1361443cded697255995fade24838ef8&to=3135761f2c94d10ee10624e11289793345d0d7ce&stat=instructions%3Au

Doesn't seem like its an issue in practice, although I agree there is a fair amount of wasted work.

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp [hunk 2]
```diff
@@ -1591,10 +1591,150 @@ static void hoistLockstepIdenticalDbgVariableRecords(
...
+      case Attribute::NonNull:
+      case Attribute::OptimizeForSize:
+        // TODO: We could merge ReadNone + Readonly -> ReadOnly
+      case Attribute::ReadNone:
+      case Attribute::ReadOnly:
+      case Attribute::Returned:
+      case Attribute::Speculatable:
+      case Attribute::WillReturn:
+      case Attribute::Writable:
+      case Attribute::WriteOnly:
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/SimplifyCFG.cpp:1664]: We should represent this as an AttrProperty in Attributes.td instead. For example as a `CanDrop` property. This way people will think about it when adding new attributes. (I think it would be best to do this as a separate PR.)

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp [hunk 3]
```diff
@@ -1591,10 +1591,150 @@ static void hoistLockstepIdenticalDbgVariableRecords(
   }
 }
 
+// See if we can intersect the attributes for two callbases (used for
+// hoisting/sinking).
+static std::optional<AttributeList> tryIntersectAttrs(const CallBase *CB0,
+                                                      const CallBase *CB1) {
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/SimplifyCFG.cpp:1597]: This intersection logic should really live in the Attributes implementation, not SimplifyCFG.

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp (hunk 2 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/Utils/SimplifyCFG.cpp:1664]: I was looking at doing that originally, decided with the lazy approach. Will do.

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp (hunk 3 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/Utils/SimplifyCFG.cpp:1597]: Thats what I originally had done, the reason I changed it was the sink/hoist intersection isn't necessarily complete as there may be attributes that change the nature of a function to a degree that makes sinking/hoisting no profitable i.e you can imagine something like the alloc attributes which are probably best not to drop.

Comment (nikic) [reply llvm/lib/Transforms/Utils/SimplifyCFG.cpp:1597]: I don't think we are interested in modelling that, but if we really wanted, we could pass a profitability callback.

Comment (goldsteinn) [reply llvm/lib/Transforms/Utils/SimplifyCFG.cpp:1597]: Likewise all the `opt*` attrs. I'll move to attributes but Ill give the API a whitelist of attrs may intersect to avoid these cases.

Comment (nikic) [reply llvm/lib/Transforms/Utils/SimplifyCFG.cpp:1597]: It's a bit borderline, but I'd not consider something like optnone a droppable attribute. Though in any case, these are not relevant to call-sites.

Comment (goldsteinn) [reply llvm/lib/Transforms/Utils/SimplifyCFG.cpp:1597]: Agreed these aren't really relevant to a call-site, but there is nothing correctness related to dropping `optnone` is there? So AFAICT the only argument against `optnone` being dropable is related to "would we want to?" which is based on the usage, hence having the code in simplifycfg where we can actually answer that question. But like you said, its not too difficult to have add a profitability check in the API.

Comment (goldsteinn) [reply llvm/lib/Transforms/Utils/SimplifyCFG.cpp:1597]: PR at https://github.com/llvm/llvm-project/pull/109719

**Comment by goldsteinn** [comment]
Rebased using new Attributes api.

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp [hunk 4]
```diff
@@ -1747,6 +1727,36 @@ static bool isSafeCheapLoadStore(const Instruction *I,
...
+                                                        const CallBase *CB1) {
+  AttributeList AL0 = CB0->getAttributes();
+  if (CB0 == CB1)
+    return AL0;
+
+  return AL0.intersectWith(CB0->getContext(), CB1->getAttributes());
+}
+
+static bool areIdenticalUpToCommutativity(const Instruction *I1,
+                                          const Instruction *I2) {
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/SimplifyCFG.cpp:1740]: Why was this function moved?

File: llvm/include/llvm/IR/Instruction.h [hunk 5]
```diff
@@ -881,16 +881,19 @@ class Instruction : public User,
   /// This is like isIdenticalTo, except that it ignores the
   /// SubclassOptionalData flags, which may specify conditions under which the
   /// instruction's result is undefined.
-  bool isIdenticalToWhenDefined(const Instruction *I) const LLVM_READONLY;
+  bool isIdenticalToWhenDefined(const Instruction *I,
+                                bool IgnoreAttrs = false) const LLVM_READONLY;
```
Comment (nikic) [inline llvm/include/llvm/IR/Instruction.h:885]: I think it would be more useful if this flag controlled whether the comparison is done strictly or via intersection.

(I was initially going to suggest that isIdenticalToWhenDefined should unconditionally use intersection as part of its "when defined" semantics and andIRFlags should handle call attribute intersection. And while that may be the best end state, I do think phasing this in as a flag first is preferable.)

File: llvm/test/Transforms/SimplifyCFG/hoist-cb-diff-attrs.ll [hunk 6]
```diff
@@ -0,0 +1,195 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --function-signature --check-attributes
+; RUN: opt < %s -passes='simplifycfg<hoist-common-insts>' -simplifycfg-require-and-preserve-domtree=1 -S | FileCheck %s
+
+target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
```
Comment (nikic) [inline llvm/test/Transforms/SimplifyCFG/hoist-cb-diff-attrs.ll:4]: Does this need data layout?

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp (hunk 4 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/Utils/SimplifyCFG.cpp:1740]: Sorry merging error on my part.

File: llvm/include/llvm/IR/Instruction.h (hunk 5 above)
Comment (goldsteinn) [reply llvm/include/llvm/IR/Instruction.h:885]: Agreed, Will change

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp [hunk 7]
```diff
@@ -1910,6 +1910,14 @@ bool SimplifyCFGOpt::hoistCommonCodeFromSuccessors(Instruction *TI,
           if (!I2->use_empty())
             I2->replaceAllUsesWith(I1);
           I1->andIRFlags(I2);
+          if (auto *CB = dyn_cast<CallBase>(I1)) {
+            bool Success = CB->tryIntersectAttributes(cast<CallBase>(I2));
+            assert(Success && "We should not be trying to hoist callbases "
+                              "with non-intersectable attributes");
+            // For NDEBUG Compile.
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/SimplifyCFG.cpp:1917]: Using `[[maybe_unused]]` may be cleaner.

File: llvm/test/Transforms/SimplifyCFG/hoist-cb-diff-attrs.ll [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+; CHECK-NEXT:    br i1 [[C]], label [[COMMON_RET:%.*]], label [[ELSE:%.*]]
+; CHECK:       common.ret:
+; CHECK-NEXT:    ret ptr [[R]]
+; CHECK:       else:
+; CHECK-NEXT:    call void @side.effect()
+; CHECK-NEXT:    br label [[COMMON_RET]]
+;
+  br i1 %c, label %if, label %else
+if:
+  %r = call ptr @foo(ptr dereferenceable(50) align 64 dereferenceable_or_null(100) %p, i64 range(i64 10, 1000) %x) memory(read)
```
Comment (nikic) [inline llvm/test/Transforms/SimplifyCFG/hoist-cb-diff-attrs.ll:20]: Having both dereferenceable and dereferenceable_or_null on the same argument is pretty weird. I'd use a separate argument.

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/test/Transforms/SimplifyCFG/hoist-cb-diff-attrs.ll (hunk 8 above)
Comment (goldsteinn) [reply llvm/test/Transforms/SimplifyCFG/hoist-cb-diff-attrs.ll:20]: Fair enough. Realize, we could actually intersect `dereferenceable` with `derefenceable_or_null` -> `derefereanceable_or_null`.

Might make a PR for it.
